Add situational conditions param

ahuang11 opened the pull request:
Partially addresses https://github.com/holoviz/lumen/issues/1161

Cleans up the Planner prompt significantly and improves separation of concerns by managing the fuzzy conditions under each agent.

Then also tweaks other prompts

Next, I want to add hard dependencies: `exclusions` (keys that are incompatible) and `not_with` (actors that are incompatible)

philippjfr commented on lumen/ai/agents.py:
```diff
@@ -504,14 +521,30 @@ async def retry_invoke(event: param.parameterized.Event):
+            "Better choice than TableListAgent when asked to show table data",
+            "Works with IterativeTableLookup when additional columns are needed",
+            "Only use for new data analysis; for visualization of existing results, use current data",
+            "Defer to DbtslAgent when dbtsl_metaset is available",
```
Only comment on this PR is that I think the more generic agent shouldn't be listing conditions about a more specific agent, i.e. I'd invert this and have DbtAgent have a condition that says: "Prefer DbtslAgent if dbtsl_metaset" is available.

philippjfr reviewed (commented):
Overall this is a great improvement. Just left one comment that I think is important for this to be scalable, i.e. generic agents can't list every possible interaction with more specific agents.

ahuang11 commented:
>  i.e. generic agents can't list every possible interaction with more specific agents.

I agree with this to a certain degree, but I think it can be very helpful for some agents to specify its relationship (i.e. SQLAgent with IterativeTableLookup), or else the Planner stumbles.

```
    conditions = param.List(default=[], doc="""
        Specific criteria that determine when this actor should be invoked.
        These conditions **establish explicit relationships between actors in the system**,
        defining the circumstances under which this actor becomes relevant.
        While 'purpose' describes what the actor does, conditions specify
        the precise situations that warrant its use.""")
```

Also, `not_with` now handles: `Defer to DbtslAgent when dbtsl_metaset is available` without mentioning it.

Other than that...
1. I've revamped most Actors' prompts
2. I've created a `_pre_plan` method in Coordinator. I realized that the exclusions must happen AFTER the planning tools are executed are outdated. For example, on the first run, `dbtsl_metaset` is generated in memory. Second run, if exclusions happen before the planning tools and DbtslLookup decides no relevant metrics, both SQLAgent and dbtsl_metaset is dropped (thus DbtslAgent), leaving no options to perform queries until the next message, but if exclusions happen after the planning tools, then the planning tools can inform what exclusions to make.
3. `is_follow_up` is now like the previous invalidation step that we removed, popping data/pipeline
4. I migrated some current memory context into `follow_up` prompt vs in the `planner` prompt

changed parts of them into conditions/exclusions/not_with and a few bugs like Actor not defining its main prompt.

philippjfr commented:
>I agree with this to a certain degree, but I think it can be very helpful for some agents to specify its relationship (i.e. SQLAgent with IterativeTableLookup), or else the Planner stumbles.

I agree, I'd think of those as a pair and very foundational.

philippjfr commented on lumen/ai/actor.py:
```diff
@@ -11,6 +11,8 @@
 from panel.chat import ChatFeed
 from pydantic import BaseModel
 
+from lumen.ai.config import PROMPTS_DIR
```
Please be consistent in the imports. I know relative imports aren't everyone's cup of tea but as long as we're using them let's be consistent about it.

ahuang11 replied:
Turns out we were very inconsistent about it heh
https://github.com/holoviz/lumen/pull/1187/commits/2edc698cc8fe11fb85a1d6c5cb8559625c4a6f7a

philippjfr reviewed (approved):
Looks good, thanks.
